File: Pastura/PasturaTests/Engine/ScoringLogic/EventReactivePayoffLogicTests.swift
import Foundation
import Testing

@testable import Pastura

/// Unit tests for `EventReactivePayoffLogic` (#931).
///
/// The favored action is read from `state.variables[favoredVariable]`;
/// these tests write it directly rather than routing through
/// `EventInjectHandler`, keeping the scoring logic under isolated test.
@Suite(.timeLimit(.minutes(1)))
struct EventReactivePayoffLogicTests {
  let logic = EventReactivePayoffLogic()
  let favoredKey = "current_event__favors"

  private func output(action: String) -> TurnOutput {
    TurnOutput(fields: ["action": action])
  }

  // MARK: - Matched / unmatched action

  @Test func rewardsOnlyAgentsWhoMatchedFavoredAction() {
    var state = SimulationState(scores: ["Alice": 0, "Bob": 0])
    state.lastOutputs = [
      "Alice": output(action: "betray"),
      "Bob": output(action: "cooperate")
    ]
    state.variables[favoredKey] = "betray"
    let collector = EventCollector()

    logic.calculate(state: &state, favoredVariable: favoredKey, emitter: collector.emit)

    // Only the betray reader gains points; the cooperate agent gains 0.
    #expect(state.scores["Alice"] == EventReactivePayoffLogic.matchReward)
    #expect(state.scores["Bob"] == 0)
  }

  @Test func symmetricForCooperateFavored() {
    var state = SimulationState(scores: ["Alice": 0, "Bob": 0])
    state.lastOutputs = [
      "Alice": output(action: "betray"),
      "Bob": output(action: "cooperate")
    ]
    state.variables[favoredKey] = "cooperate"
    let collector = EventCollector()

    logic.calculate(state: &state, favoredVariable: favoredKey, emitter: collector.emit)

    #expect(state.scores["Alice"] == 0)
    #expect(state.scores["Bob"] == EventReactivePayoffLogic.matchReward)
  }

  @Test func rewardAddsToExistingScore() {
    var state = SimulationState(scores: ["Alice": 5])
    state.lastOutputs = ["Alice": output(action: "betray")]
    state.variables[favoredKey] = "betray"
    let collector = EventCollector()

    logic.calculate(state: &state, favoredVariable: favoredKey, emitter: collector.emit)

    #expect(state.scores["Alice"] == 5 + EventReactivePayoffLogic.matchReward)
  }

  // MARK: - No-op paths (missing / empty favored var, plain-string list)

  @Test func missingFavoredVariableIsNoOp() {
    var state = SimulationState(scores: ["Alice": 0, "Bob": 0])
    state.lastOutputs = [
      "Alice": output(action: "betray"),
      "Bob": output(action: "cooperate")
    ]
    // favoredKey absent → simulates a plain-string event list (no companion var).
    let collector = EventCollector()

    logic.calculate(state: &state, favoredVariable: favoredKey, emitter: collector.emit)

    #expect(state.scores["Alice"] == 0)
    #expect(state.scores["Bob"] == 0)
  }

  @Test func emptyFavoredVariableIsNoOp() {
    var state = SimulationState(scores: ["Alice": 0])
    state.lastOutputs = ["Alice": output(action: "betray")]
    // "" is what EventInjectHandler writes on a miss round — must not reward.
    state.variables[favoredKey] = ""
    let collector = EventCollector()

    logic.calculate(state: &state, favoredVariable: favoredKey, emitter: collector.emit)

    #expect(state.scores["Alice"] == 0)
  }

  @Test func alwaysEmitsScoreUpdate() {
    var state = SimulationState(scores: ["Alice": 0])
    state.lastOutputs = ["Alice": output(action: "betray")]
    // No favored var → inert, but a scoreUpdate must still be emitted so the
    // UI/persistence pipeline sees a well-formed code-phase result.
    let collector = EventCollector()

    logic.calculate(state: &state, favoredVariable: favoredKey, emitter: collector.emit)

    let scoreUpdates = collector.events.filter {
      if case .scoreUpdate = $0 { return true }
      return false
    }
    #expect(scoreUpdates.count == 1)
  }

  // MARK: - Agent gating

  @Test func ignoresOutputsWithoutASeededScore() {
    var state = SimulationState(scores: ["Alice": 0])
    state.lastOutputs = [
      "Alice": output(action: "betray"),
      // No score seeded for "Ghost" → gated out (not a live agent).
      "Ghost": output(action: "betray")
    ]
    state.variables[favoredKey] = "betray"
    let collector = EventCollector()

    logic.calculate(state: &state, favoredVariable: favoredKey, emitter: collector.emit)

    #expect(state.scores["Alice"] == EventReactivePayoffLogic.matchReward)
    #expect(state.scores["Ghost"] == nil)
  }

  // MARK: - Action normalization (raw LLM output is un-canonicalized)

  @Test func rewardsCaseAndWhitespaceVariantActions() {
    // The individual `choose` path stores the raw LLM `action` and the value
    // is not grammar-constrained, so casing/whitespace variants must still
    // count as a match against a lowercase curated `favors`.
    var state = SimulationState(scores: ["Alice": 0, "Bob": 0, "Carol": 0])
    state.lastOutputs = [
      "Alice": output(action: "Betray"),
      "Bob": output(action: " betray "),
      "Carol": output(action: "BETRAY")
    ]
    state.variables[favoredKey] = "betray"
    let collector = EventCollector()

    logic.calculate(state: &state, favoredVariable: favoredKey, emitter: collector.emit)

    #expect(state.scores["Alice"] == EventReactivePayoffLogic.matchReward)
    #expect(state.scores["Bob"] == EventReactivePayoffLogic.matchReward)
    #expect(state.scores["Carol"] == EventReactivePayoffLogic.matchReward)
  }

  @Test func doesNotRewardNonMatchingActionAfterNormalization() {
    var state = SimulationState(scores: ["Alice": 0])
    state.lastOutputs = ["Alice": output(action: "Cooperate")]
    state.variables[favoredKey] = "betray"
    let collector = EventCollector()

    logic.calculate(state: &state, favoredVariable: favoredKey, emitter: collector.emit)

    #expect(state.scores["Alice"] == 0)
  }

  // MARK: - v1 custom-`as:` deferral boundary (ScoreCalcHandler dispatch)

  @Test func customEventVariableScoresNothingUnderV1() async throws {
    // ScoreCalcHandler v1 hardwires the default `current_event__favors` key.
    // A dict event under a custom `as:` writes `biz_event__favors`, so
    // `event_reactive` scores nothing until the deferred `favored_variable:`
    // wiring lands (#931). Pin the boundary so that future change is a
    // visible diff, not a silent behavior shift.
    let scenario = makeTestScenario(
      agentNames: ["Alice"],
      phases: [Phase(type: .scoreCalc, logic: .eventReactive)]
    )
    var state = SimulationState.initial(for: scenario)
    // Favored action lives under the custom key, NOT the default one.
    state.variables["biz_event__favors"] = "betray"
    state.lastOutputs = ["Alice": output(action: "betray")]
    let mock = MockLLMService(responses: [])
    let collector = EventCollector()
    let context = makePhaseContext(scenario: scenario, llm: mock, collector: collector)

    try await ScoreCalcHandler().execute(context: context, state: &state)

    #expect(state.scores["Alice"] == 0)
  }
}

File: shared/engine/src/commonTest/kotlin/com/pastura/engine/ScoringLogic/EventReactivePayoffLogicTests.kt
package com.pastura.engine

import com.pastura.models.SimulationEvent
import com.pastura.models.SimulationState
import com.pastura.models.TurnOutput
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

/**
 * Kotlin sibling of Swift's `EventReactivePayoffLogicTests` (#931).
 *
 * The favored action is read from `state.variables[favoredVariable]`; these
 * tests write it directly rather than routing through `EventInjectHandler`.
 * (Swift's `customEventVariableScoresNothingUnderV1` case is a ScoreCalcHandler
 * dispatch test — that handler is not part of this slice — so it is omitted.)
 *
 * Ported for the ADR-023 Stage-3 PR-1 score_calc slice (#501).
 */
class EventReactivePayoffLogicTests {

    private val logic = EventReactivePayoffLogic()
    private val favoredKey = "current_event__favors"
    private val reward = EventReactivePayoffLogic.matchReward

    private fun output(action: String) = TurnOutput(fields = mapOf("action" to action))

    @Test
    fun rewardsOnlyAgentsWhoMatchedFavoredAction() {
        val state = SimulationState(
            scores = mapOf("Alice" to 0, "Bob" to 0),
            lastOutputs = mapOf("Alice" to output("betray"), "Bob" to output("cooperate")),
            variables = mapOf(favoredKey to "betray"),
        )
        val next = logic.calculate(state, favoredKey) { }
        assertEquals(reward, next.scores["Alice"])
        assertEquals(0, next.scores["Bob"])
    }

    @Test
    fun symmetricForCooperateFavored() {
        val state = SimulationState(
            scores = mapOf("Alice" to 0, "Bob" to 0),
            lastOutputs = mapOf("Alice" to output("betray"), "Bob" to output("cooperate")),
            variables = mapOf(favoredKey to "cooperate"),
        )
        val next = logic.calculate(state, favoredKey) { }
        assertEquals(0, next.scores["Alice"])
        assertEquals(reward, next.scores["Bob"])
    }

    @Test
    fun rewardAddsToExistingScore() {
        val state = SimulationState(
            scores = mapOf("Alice" to 5),
            lastOutputs = mapOf("Alice" to output("betray")),
            variables = mapOf(favoredKey to "betray"),
        )
        val next = logic.calculate(state, favoredKey) { }
        assertEquals(5 + reward, next.scores["Alice"])
    }

    @Test
    fun missingFavoredVariableIsNoOp() {
        // favoredKey absent → simulates a plain-string event list (no companion var).
        val state = SimulationState(
            scores = mapOf("Alice" to 0, "Bob" to 0),
            lastOutputs = mapOf("Alice" to output("betray"), "Bob" to output("cooperate")),
        )
        val next = logic.calculate(state, favoredKey) { }
        assertEquals(0, next.scores["Alice"])
        assertEquals(0, next.scores["Bob"])
    }

    @Test
    fun emptyFavoredVariableIsNoOp() {
        // "" is what EventInjectHandler writes on a miss round — must not reward.
        val state = SimulationState(
            scores = mapOf("Alice" to 0),
            lastOutputs = mapOf("Alice" to output("betray")),
            variables = mapOf(favoredKey to ""),
        )
        val next = logic.calculate(state, favoredKey) { }
        assertEquals(0, next.scores["Alice"])
    }

    @Test
    fun alwaysEmitsScoreUpdate() {
        // No favored var → inert, but a scoreUpdate must still be emitted.
        val state = SimulationState(
            scores = mapOf("Alice" to 0),
            lastOutputs = mapOf("Alice" to output("betray")),
        )
        val events = mutableListOf<SimulationEvent>()
        logic.calculate(state, favoredKey) { events += it }
        assertEquals(1, events.filterIsInstance<SimulationEvent.ScoreUpdate>().size)
    }

    @Test
    fun ignoresOutputsWithoutASeededScore() {
        val state = SimulationState(
            scores = mapOf("Alice" to 0),
            // No score seeded for "Ghost" → gated out (not a live agent).
            lastOutputs = mapOf("Alice" to output("betray"), "Ghost" to output("betray")),
            variables = mapOf(favoredKey to "betray"),
        )
        val next = logic.calculate(state, favoredKey) { }
        assertEquals(reward, next.scores["Alice"])
        assertNull(next.scores["Ghost"])
    }

    @Test
    fun rewardsCaseAndWhitespaceVariantActions() {
        // Casing/whitespace variants must still count against a lowercase favors.
        val state = SimulationState(
            scores = mapOf("Alice" to 0, "Bob" to 0, "Carol" to 0),
            lastOutputs = mapOf(
                "Alice" to output("Betray"),
                "Bob" to output(" betray "),
                "Carol" to output("BETRAY"),
            ),
            variables = mapOf(favoredKey to "betray"),
        )
        val next = logic.calculate(state, favoredKey) { }
        assertEquals(reward, next.scores["Alice"])
        assertEquals(reward, next.scores["Bob"])
        assertEquals(reward, next.scores["Carol"])
    }

    @Test
    fun doesNotRewardNonMatchingActionAfterNormalization() {
        val state = SimulationState(
            scores = mapOf("Alice" to 0),
            lastOutputs = mapOf("Alice" to output("Cooperate")),
            variables = mapOf(favoredKey to "betray"),
        )
        val next = logic.calculate(state, favoredKey) { }
        assertEquals(0, next.scores["Alice"])
    }
}
